sample-apps/celery_sample/Procfile

```diff
@@ -0,0 +1,7 @@
+# The proxy server adds the X_REQUEST_START header for queue time calculations
+proxy: npx judoscale-adapter-proxy-server
+
+web: DYNO=web.1 poetry run gunicorn --preload 'app:create_app()'
+worker: DYNO=worker.1 poetry run celery -A app worker -c 1 --loglevel=INFO -Q low
+worker_high: DYNO=worker_high.1 poetry run celery -A app worker -c 1 --loglevel=INFO -Q high
```

Comment:
I was expecting to see the reporter running on all three processes, but it appears to only be running on `web.1`:

```
8:36:50 AM web.1         |  DEBUG - [Judoscale] Collecting metrics for queues ['high', 'low']
8:36:50 AM web.1         |  DEBUG - [Judoscale] queue_name=high, queue_time=10567ms
8:36:50 AM web.1         |  DEBUG - [Judoscale] queue_name=low, queue_time=10587ms
8:36:50 AM web.1         |  DEBUG - [Judoscale] Posting 10 metrics from 39677 to Judoscale adapter API https://requestinspector.com/inspect/judoscale-celery/v3/reports
8:36:53 AM web.1         |  DEBUG - [Judoscale] Collecting metrics for queues []
8:36:53 AM web.1         |  DEBUG - [Judoscale] Posting 0 metrics from 39487 to Judoscale adapter API https://requestinspector.com/inspect/judoscale-celery/v3/reports
8:37:06 AM web.1         |  DEBUG - [Judoscale] Collecting metrics for queues ['high', 'low']
8:37:06 AM web.1         |  DEBUG - [Judoscale] queue_name=high, queue_time=25681ms
8:37:06 AM web.1         |  DEBUG - [Judoscale] queue_name=low, queue_time=25808ms
8:37:06 AM web.1         |  DEBUG - [Judoscale] Posting 2 metrics from 39677 to Judoscale adapter API https://requestinspector.com/inspect/judoscale-celery/v3/reports
8:37:09 AM web.1         |  DEBUG - [Judoscale] Collecting metrics for queues []
8:37:09 AM web.1         |  DEBUG - [Judoscale] Posting 0 metrics from 39487 to Judoscale adapter API https://requestinspector.com/inspect/judoscale-celery/v3/reports
```

Will this still work if a user is running a "worker-only" app with no web process?

I'm also confused why only one of the two Gunicorn processes is reporting the Celery queues. Functionally that's fine since we only need one set of worker metrics, but can you help me understand why that's happening?

Reply:
> Will this still work if a user is running a "worker-only" app with no web process?

Hmm.. no. It seems I didn't think of this use case. So web dynos can also scale down to 0? Can web and worker dynos be scaled down to 0 at the same time?

> I'm also confused why only one of the two Gunicorn processes is reporting the Celery queues. Functionally that's fine since we only need one set of worker metrics, but can you help me understand why that's happening?

Yep — Gunicorn starts a master process and N child processes to handle web requests. The master process itself will never see any web requests, it only supervises the child processes. The master process does not see the Celery queues because the collector gets the queue names from the `before_task_publish` handler, which is executed in the same process from which `.apply_async` is called — e.g the child process handling web requests.

Essentially, the reporter and the collectors exist in separate processes and are isolated from each other. The master process will never see metrics because it'll never handle any web requests and thus never enqueue any tasks.

Reply:
We can make the worker process also start a reporter, that's not a huge obstacle. The larger issue is how the reporter in the worker process will know which queues to report on, if we only know the queue name in the signal handler in the process we call `.apply_async()` in...

If the reporter has to run in the worker process as well then I think the current approach with getting queue names from the signal handler won't work.

Reply:
I think we can make this work by using the [`task-sent` event](https://docs.celeryq.dev/en/latest/userguide/monitoring.html#task-sent) and a slightly more complicated metrics collector, which needs to handle the `task-sent` event as well.

I'll prototype it to see if there are any obvious roadblocks.

Reply:
@adamlogic check out the latest changes. All three processes (web and two workers) are running the reporter now and reporting metrics as I think they should.